pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -253,18 +253,18 @@ class ModelInterpreter implements Serializable {
                         // These are literals that guaranteed not to depend on other variables
                         // Users will expect them to be available at all times in a particular cell in a matrix.
                         withEnvBlock(thisStage.getMatrixCellEnvVars(script)) {
-                            inWrappers(thisStage.options?.wrappers) {
+                            boolean whenBeforeOptions = thisStage.when?.beforeOptions ?: false
+                            boolean whenPassed = whenBeforeOptions && evaluateWhen(thisStage.when)
+                            (!whenBeforeOptions || whenPassed) && inWrappers(thisStage.options?.wrappers) {
```

Comment:
Note: Here I tried hard not to introduce another nesting level via another `if`.
Maybe not ideal, but another nesting level would be much less "readable" that this, IMHO.

Reply:
There has to be a better way to do this.  

I'm having a very hard time understanding this. What about adding a `WhenEvaluator` class and ending up with something list this:
```java
withEnvBlock(thisStage.getMatrixCellEnvVars(script)) {

    def whenEvaluator = new WhenEvaluator(thisStage.when)

    if (whenEvaluator.passedOrNotEvaluatedBeforeOptions()) {
        inWrappers(thisStage.options?.wrappers) {
            if (whenEvaluator.passedOrNotEvaluatedBeforeInput()) {
                stageInput(thisStage.input) {
                    if (thisStage?.parallel != null) {
                        if (whenEvaluator.passedOrNotEvaluated()) {
                            withCredentialsBlock(thisStage.environment) {
                                withEnvBlock(thisStage.getEnvVars(script)) {
                                    script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, null))
                                }
                            }
                        }
                    } else {
                        if (whenEvaluator.passedOrNotEvaluatedBeforeAgent()) {
                            inDeclarativeAgent(thisStage, root, thisStage.agent) {
                                if (whenEvaluator.passedOrNotEvaluated()) {
                                    withCredentialsBlock(thisStage.environment) {
                                        withEnvBlock(thisStage.getEnvVars(script)) {
                                            toolsBlock(thisStage.tools, thisStage.agent ?: root.agent, parent?.tools ?: root.tools) {
                                                if (thisStage?.stages) {
                                                    def nestedError = evaluateSequentialStages(root, thisStage.stages, firstError, thisStage, null, null).call()

                                                    // Propagate any possible error from the sequential stages as if it were an error thrown directly.
                                                    if (nestedError != null) {
                                                        throw nestedError
                                                    }
                                                } else {
                                                    // Execute the actual stage and potential post-stage actions
                                                    executeSingleStage(root, thisStage, parentAgent)
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if (!whenEvaluator.passed) {
        skippedReason = new SkippedStageReason.When(thisStage.name)
        skipStage(root, parentAgent, thisStage, firstError, skippedReason, parent).call()
    }
}
```

With some minor modifications you could end up with this: 
```
withEnvBlock(thisStage.getMatrixCellEnvVars(script)) {

    def whenEvaluator = new WhenEvaluator(thisStage.when)

    whenEvaluator.inWrappers(thisStage.options?.wrappers) {
        whenEvaluator.stageInput(thisStage.input) {
        if (thisStage?.parallel != null) {
                whenEvaluator.withCredentialsBlock(thisStage.environment) {
                    withEnvBlock(thisStage.getEnvVars(script)) {
                        script.parallel(getParallelStages(root, parentAgent, thisStage, firstError, null))
                    }
                }
            } else {
                whenEvaluator.inDeclarativeAgent(thisStage, root, thisStage.agent) {
                    whenEvaluator.withCredentialsBlock(thisStage.environment) {
                        withEnvBlock(thisStage.getEnvVars(script)) {
                            toolsBlock(thisStage.tools, thisStage.agent ?: root.agent, parent?.tools ?: root.tools) {
                                if (thisStage?.stages) {
                                    def nestedError = evaluateSequentialStages(root, thisStage.stages, firstError, thisStage, null, null).call()

                                    // Propagate any possible error from the sequential stages as if it were an error thrown directly.
                                    if (nestedError != null) {
                                        throw nestedError
                                    }
                                } else {
                                    // Execute the actual stage and potential post-stage actions
                                    executeSingleStage(root, thisStage, parentAgent)
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if (!whenEvaluator.passed) {
        skippedReason = new SkippedStageReason.When(thisStage.name)
        skipStage(root, parentAgent, thisStage, firstError, skippedReason, parent).call()
    }
}

Reply:
`WhenEvaluator` is a good idea! I'll have a look at your first suggestion.
The second one doesn't seem so minor to me and it seems like a general refactoring which is a bit out of scope for JENKINS-51865, but I might be wrong.

PS: Do you have any idea why `ModelInterpreter.groovy` is located within `src/main/resources` and _not_ `src/main/groovy`?

Reply:
> PS: Do you have any idea why `ModelInterpreter.groovy` is located within `src/main/resources` and _not_ `src/main/groovy`?

Nevermind, I learned the hard way that it needs to be there because of some CPS voodoo.
I'll push an update soon...

Reply:
Ok, update is in.
I lost too much time tackling this CPS problem (I had `WhenEvaluator` in `src/main/groovy` which silently killed `when.rawClosure.call()`), so I won't be able to spend any more time on your second suggestion, sorry.